oxidation/libparsec/crates/core/src/invite/greeter.rs

```diff
@@ -0,0 +1,669 @@
+            .map_err(|e| InviteError::Custom(e.to_string()))?;
+
+        Ok(DeviceGreetInProgress4Ctx {
+            token: self.0.token,
+            requested_device_label: data.requested_device_label,
+            verify_key: data.verify_key,
+            shared_secret_key: self.0.shared_secret_key,
+            cmds: self.0.cmds,
+        })
+    }
+}
+
+/// Helper to prepare the creation of a new user.
+fn create_new_user_certificates(
+    author: &LocalDevice,
+    device_label: Option<DeviceLabel>,
+    human_handle: Option<HumanHandle>,
+    profile: UserProfile,
+    public_key: PublicKey,
+    verify_key: VerifyKey,
+) -> (Vec<u8>, Vec<u8>, Vec<u8>, Vec<u8>, InviteUserConfirmation) {
+    let device_id = DeviceID::default();
+    let timestamp = author.now();
+
+    let user_certificate = UserCertificate {
+        author: CertificateSignerOwned::User(author.device_id.clone()),
+        timestamp,
+        user_id: device_id.user_id().clone(),
+        human_handle: human_handle.clone(),
+        public_key: public_key.clone(),
+        profile,
+    };
+
+    let redacted_user_certificate = UserCertificate {
+        author: CertificateSignerOwned::User(author.device_id.clone()),
+        timestamp,
+        user_id: device_id.user_id().clone(),
+        human_handle: None,
+        public_key,
+        profile,
+    };
+
+    let device_certificate = DeviceCertificate {
+        author: CertificateSignerOwned::User(author.device_id.clone()),
+        timestamp,
+        device_id: device_id.clone(),
+        device_label: device_label.clone(),
+        verify_key: verify_key.clone(),
+    };
+
+    let redacted_device_certificate = DeviceCertificate {
+        author: CertificateSignerOwned::User(author.device_id.clone()),
+        timestamp,
+        device_id: device_id.clone(),
+        device_label: None,
+        verify_key,
+    };
+
+    let user_certificate_bytes = user_certificate.dump_and_sign(&author.signing_key);
+    let redacted_user_certificate_bytes =
+        redacted_user_certificate.dump_and_sign(&author.signing_key);
+    let device_certificate_bytes = device_certificate.dump_and_sign(&author.signing_key);
+    let redacted_device_certificate_bytes =
+        redacted_device_certificate.dump_and_sign(&author.signing_key);
+
+    let invite_user_confirmation = InviteUserConfirmation {
+        device_id,
+        device_label,
+        human_handle,
+        profile,
+        root_verify_key: author.root_verify_key().clone(),
+    };
+
+    (
+        user_certificate_bytes,
+        redacted_user_certificate_bytes,
+        device_certificate_bytes,
+        redacted_device_certificate_bytes,
+        invite_user_confirmation,
+    )
+}
+
+// GreetInProgress4Ctx
+
+#[derive(Debug)]
+pub struct UserGreetInProgress4Ctx {
+    pub token: InvitationToken,
+    pub requested_device_label: Option<DeviceLabel>,
+    pub requested_human_handle: Option<HumanHandle>,
```

Comment:
Is it normal that those are option, Is it because it's the device & human name that we would like to have and are set by the Greeter ?

Reply:
@touilleMan

Reply:
Device label & human handle are not mandatory (this is mostly for legacy reason)
The usecase for this now is to consider a user/device can represent a non-human system

Reply:
How can you identify a `non-human` system if it don't have a `human_handle` or `device_label` ?

Reply:
@touilleMan